src/stack_limiter/mod.rs

```diff
@@ -296,85 +303,42 @@ fn instrument_function(ctx: &mut Context, func: &mut Instructions) -> Result<(),
 		};
 
 		if did_instrument {
-			calls.next();
+			call_peeker.next();
 		} else {
-			new_instrs.push(instr);
+			func_code_builder.instruction(&DefaultTranslator.translate_op(&instr)?);
 		}
 	}
 
-	if calls.next().is_some() {
-		return Err("Not all calls were used")
+	if call_peeker.next().is_some() {
+		return Err(anyhow!("not all calls were used"))
 	}
 
-	Ok(())
-}
-
-fn resolve_func_type(
-	func_idx: u32,
-	module: &elements::Module,
-) -> Result<&elements::FunctionType, &'static str> {
-	let types = module.type_section().map(|ts| ts.types()).unwrap_or(&[]);
-	let functions = module.function_section().map(|fs| fs.entries()).unwrap_or(&[]);
-
-	let func_imports = module.import_count(elements::ImportCountType::Function);
-	let sig_idx = if func_idx < func_imports as u32 {
-		module
-			.import_section()
-			.expect("function import count is not zero; import section must exists; qed")
-			.entries()
-			.iter()
-			.filter_map(|entry| match entry.external() {
-				elements::External::Function(idx) => Some(*idx),
-				_ => None,
-			})
-			.nth(func_idx as usize)
-			.expect(
-				"func_idx is less than function imports count;
-				nth function import must be `Some`;
-				qed",
-			)
-	} else {
-		functions
-			.get(func_idx as usize - func_imports)
-			.ok_or("Function at the specified index is not defined")?
-			.type_ref()
-	};
-	let Type::Function(ty) = types
-		.get(sig_idx as usize)
-		.ok_or("The signature as specified by a function isn't defined")?;
-	Ok(ty)
+	Ok(func_code_builder)
 }
 
 #[cfg(test)]
 mod tests {
 	use super::*;
-	use parity_wasm::elements;
-
-	fn parse_wat(source: &str) -> elements::Module {
-		elements::deserialize_buffer(&wat::parse_str(source).expect("Failed to wat2wasm"))
-			.expect("Failed to deserialize the module")
-	}
 
-	fn validate_module(module: elements::Module) {
-		let binary = elements::serialize(module).expect("Failed to serialize");
-		wasmparser::validate(&binary).expect("Invalid module");
+	fn parse_wat(source: &str) -> ModuleInfo {
+		let module_bytes = wat::parse_str(source).unwrap();
+		ModuleInfo::new(&module_bytes).unwrap()
 	}
 
 	#[test]
 	fn test_with_params_and_result() {
-		let module = parse_wat(
-			r#"
-(module
-	(func (export "i32.add") (param i32 i32) (result i32)
-		get_local 0
-	get_local 1
-	i32.add
-	)
-)
-"#,
+		let mut module_info = parse_wat(
+			r#"(module
+						(func (export "i32.add") (param i32 i32) (result i32)
+							get_local 0
+							get_local 1
+							i32.add
+						)
+					)"#,
 		);
 
-		let module = inject(module, 1024).expect("Failed to inject stack counter");
-		validate_module(module);
+		let inject_raw_wasm =
+			inject(&mut module_info, 1024).expect("Failed to inject stack counter");
+		wasmparser::validate(&inject_raw_wasm).expect("Invalid module");
```

Comment:
Can we assert the instrumented code? This will force us to triple check the results.